library/random/alphanumeric_spec.rb

```diff
@@ -0,0 +1,57 @@
+require_relative '../../spec_helper'
+
+ruby_version_is "3.1" do
+  require 'random/formatter'
```

Comment:
Looks like `Random::Formatter` was available in Ruby 3.0 but was part of `securerandom`. 

> Random::Formatter is moved to random/formatter.rb, so that you can use Random#hex, Random#base64, and so on without SecureRandom. [[Feature #18190](https://bugs.ruby-lang.org/issues/18190)]

https://github.com/ruby/ruby/blob/v3_1_0/NEWS.md
https://bugs.ruby-lang.org/issues/18190

And it seems `Random::Formatter`'s methods are available on `SecureRandom`:

```ruby
SecureRandom.method(:alphanumeric)
# => #<Method: #<Class:SecureRandom>(Random::Formatter)#alphanumeric(n=..., chars: ...)
```

So ideally we would want to have a shared example to test `#alphanumeric` for both `Random::Formatter` and `SequreRandom`.

JFYI there are specs of `Random::Formatter` methods in https://github.com/ruby/spec/tree/master/library/securerandom that would probably also might be shared.

Reply:
And it seems `Random::Formatter`'s methods are available on `SecureRandom`. So ideally we would want to have a shared example to test `#alphanumeric` for both `Random::Formatter` and `SequreRandom`.

```ruby
SecureRandom.method(:alphanumeric)
# => #<Method: #<Class:SecureRandom>(Random::Formatter)#alphanumeric(n=..., chars: ...)
```

JFYI there are specs of `Random::Formatter` methods in https://github.com/ruby/spec/tree/master/library/securerandom that would probably also might be shared.

Reply:
I think I've got the correct version now:

* Shared spec in `shared/random/formatter/alphanumeric.rb`
* Spec for `Random.alphanumeric` in Ruby 3.1 and above which depends on `Random::Formatter`
* Spec for `SecureRandom.alphanumeric`, depends on `Random::Formatter` in Ruby 3.1 and above, but runs on Ruby 3.0 as well.

If this is merged, we can use the same framework to convert a number of specs in `library/securerandom` to shared formatter specs.

Reply:
I really apologise. I have completely forgotten about the owning convention in `ruby/spec`.

A method specs are added only for a class/module where this method is defined. It doesn't make sense to test e.g. the `Enumerable#each` method for all the core classes that include the `Enumerable` module.

So in our case specs should cover only the `Random::Formatter` module and don't check `Random` and `SecureRandom` because they just include the `Random::Formatter` module:

```ruby
SecureRandom.method(:alphanumeric)
# => #<Method: #<Class:SecureRandom>(Random::Formatter)#alphanumeric(n=..., chars: ...)  ...
Random.method(:alphanumeric)
# => #<Method: #<Class:Random>(Random::Formatter)#alphanumeric(n=..., chars: ...) ...
```

Reply:
It means that existing specs for `SecureRandom`'s methods that are defined in `Random::Formatter` (e.g. `hex`, `random_bytes`, `random_number`, `base64`) are supposed to be moved to specs for `Random::Formatter`.

Reply:
I pushed a new version, this only includes the new file `library/random/formatter/alphanumeric_spec.rb` and uses a mock object (unrelated to `Random` and `SecureRandom`) to generate the random stream (XKCD 221 kind of random). This means the output is predictable, but I don't think that should have explicit tests (I would consider that more of an implementation detail).

I was planning on converting the existing specs of SecureRandom that test Random::Formatter to the new folder after this one is merged, just to make sure we have a definitive style first.

Reply:
Sounds good. Thank you!